Assets/Scripts/Input/Player.cs

```diff
@@ -0,0 +1,217 @@
+
+    /// <summary>
+    /// Interprets the movement of the player,
+    /// uses raycast to resolve movement behaviour
+    /// </summary>
+    /// <param name="xVelo">x axis velocity</param>
+    /// <param name="yVelo">y axis velocity</param>
+    private void Move(float xVelo, float yVelo)
+    {
+        //if airborne
+        if (!grounded)
+        {
+            //and rising
+            if (yVelo > 0)
+            {
+                //detect collision with above objects
+                if (CastRaysUp())
+                {
+                    cielingCollision = true;
+                }
+            }
+            //if falling, and jumpingGrace has passed
+            else if(!jumpingGrace)
+            {
+                //detect collision with the ground
+                if (CastRaysDown())
+                {
+                    Land();
+                    isLanding = true;
+                }
+            }
+        }
+
+        //if on the ground, detect falling
+        else if (!CastRaysDown())
+        {
+            grounded = false;
+        }
+
+        //resolve movement
+        transform.Translate(xVelo * Time.deltaTime, yVelo * Time.deltaTime, 0);
+
+        //post movement approximation here
+        if (isLanding)
+        {
+            ApproximateYPosition();
+            isLanding = false;
+        }
+            
+    }
+
+    private void FixedUpdate()
+    {
+        if (!grounded)
+        {
+            if (cielingCollision)
+            {
+                yVelo = -dropFromCielingVelocity;
+                cielingCollision = false;
+                return;
+            }
+            yVelo = yVelo + yAccel * Time.deltaTime;
+            yVelo = Mathf.Clamp(yVelo, -fallVelocityLimit, jumpVelocity);
+        }
+    }
+    #endregion
+
+    #region Run and jump
+
+    public void Run(float direction)
+    {
+        if (CastHoriRays(direction))
+        {
+            xVelo = 0;
+            ApproximateXPosition(direction);
+        }
+        else
+            xVelo = direction * runningVelocity;
+
+    }
+
+    public void JumpAction()
+    {
+        if (grounded)
+        {
+            yVelo += jumpVelocity;
+            grounded = false;
+            jumpingGrace = true;
+            StartCoroutine("HandleJumpGrace");
+        }
+            
+    }
+    #endregion
+
+    #region Auxiliry and approximation methods
+
+    private void Land()
+    {
+        grounded = true;
+        yVelo = 0;
+    }
+
+    IEnumerator HandleJumpGrace()
+    {
+        yield return new WaitForSeconds(jumpingGraceTime);
+        jumpingGrace = false;
+    }
+
+    private void ApproximateYPosition()
+    {
+        float yValue = Mathf.Floor(transform.position.y) + 0.5f - (2 - height)/2;
+        yValue += floatMargin;
+        transform.position = new Vector3(transform.position.x, yValue, transform.position.z);
+    }
+
+    private void ApproximateXPosition(float direction)
```

Comment:
Not sure the reason for the approximate functions but they wont work unless each block is set perfectly on whole numbers

Reply:
The blocks are on perfectly whole nubmers though. This is to place the player in a consistent place when colliding with walls and landing. It has been buggy before.

Will try placing movement in fixed update, it might prevent the inconsistency.

Reply:
It didn't, this is necessary.